doc/man/66-scandir.1.scd
66-scandir(1)

# NAME

66-scandir - Handle the scandir for a given user

# SYNOPSYS

66-scandir [ *-h* ] [ *-z* ] [ *-v* _verbosity_ ] [ *-b* ] [ *-l* _live_ ] [* -t* _rescan_ ] [ *-L* _log_user_ ] [ *-s* _skel_ ] [ *-e* _environment_ ] [ *-c* | *-r* | *-u* ] _owner_

# DESCRIPTION

*66-scandir* creates or start the scandir (a.k.a. directory containing a
collection of *s6-supervise* processes) for the given _owner_ depending on the
provided options. Note that _owner_ can be any valid user on the system. However, the given user must have sufficient permissions
to create the necessary directories at its location. That is *%%livedir%%* by
default or the resulting path provided by the *-l* option. If _owner_ is not
explicitely set then the user of the current process will be used instead.

# OPTIONS

*-h*
	Prints this help.

*-z*
	use color.

*-v* _verbosity_
	Increases/decreases the verbosity of the command.++
	*1* : (Default) Only print error messages.++
	*2* : Also print warning messages.++
	*3* : Also print tracing messages.++
	*4* : Also print debugging messages.

*-b*
	Create specific files for boot. Only root user can use this option. It is
	not meant to be used directly even with root. *66-boot*(8) calls it during
	the boot process.

*-l* _live_
	An absolute path. Create the scandir at _live_. Default is *%%livedir%%*.
	The default can also be changed at compile-time by passing the
	\--livedir=_live_ option to *./configure*. Should be under a writable
	filesystem - likely a RAM filesystem.

*-d* _notif_
	notify readiness on file descriptor _notif_. When scandir is ready to 
	accept commands from *66-scanctl*, it will write a newline to _notif_.
	_notif_ cannot be lesser than 3. By default, no notification is sent.
	if *-b* is set, this option have no effects.

*-t* _rescan_
	Perform a scan every _rescan_ milliseconds. If _rescan_ is set to *0*
	(the default), automatic scans are never performed after the first
	one and *s6-svscan* will only detect new services by  issuing either
	"66-scandir -s reload" or "s6-svscanctl -a". It is *strongly*
	discouraged to set _rescan_ to a positive value under 500.

*-s* _skel_
	An absolute path. Directory  containing _skeleton_ files. This option is
	not meant to be used directly even with root. *66-boot*(8) program calls it
	at during the boot process. Default is %%skel%%.

*-L* _log_user_
	Will run as the catch-all logger as _log_user_ user. Default is *%%s6log_user%%*.
	The default can also be changed at compile-time by passing the
	\--with-s6-log-user=_user_ option to *./configure*.

*-e* _environment_
	An absolute path. Merge the current environment variables of _scandir_
	with variables found in the directory. Any file in _environment_
	not beginning with a dot and not containing the '=' character will
	be read.

*-c*
	Create the neccessary directories at _live_.
	Must be executed once with this option before being able to use *-u*.

*-r*
	Remove the scandir directory at _live_.
	If it was already started with *-u* it must first be stopped sending
	a signal with "66-scanctl quit" or similar.

*-u*
	Start the _scandir_ directory at _live_ invocating the *s6-svscan* program.

# SCANDIR CREATION PROCESS

When creating the _scandir_ with the *-c* option various files and directories
will be created at the _live_ directory for the given _owner_.

If created with the *root* user, you will find the following in *%%livedir%%*
(the directory created by default if *-l* is not passed and *0* being the
corresponding UID for the root user):

*%%livedir%%/scandir/0* : stores all longrun proccesses (commonly known as
	daemons) started by the user *root*.
*%%livedir%%/tree/0* : stores any s6-rc service database started by *root*.
*%%livedir%%/log/0* : stores the catch-all logger when the scandir is created
	for a boot procedure with the *-b* option.
*%%livedir%%/state/0* : stores internal 66 ecosystem files.

If created with a regular user you will find the following in *%%livedir%%*
(Default directories if *-l* is not passed and *1000* being the UID for the
user):

*%%livedir%%/scandir/1000*

*%%livedir%%/tree/1000*

*%%livedir%%/state/1000*

If a _scandir_ already exists for the given user it will prevent
the creation when issuing "66-scandir -c". If you want to create a different
_scandir_ for the same _owner_ you must delete it first with *-r*.

# ENVIRONMENT CONFIGURATION

You can modify environment variables when starting the _scandir_ with the *-e*
option. This option expects to find a valid absolute path for a directory
containing one or more files where the format is the classic _key=value_ pair.
Each file found will be read and parsed. See *execl-envfile*(5) for more.

Any longrun process launched on the scandir should inherit the environment
variables of the scandir. A specific global _key=value_ pair inherited by all
daemons can be set using this option.

# SCANDIR REMOVE PROCESS

When using the *-r* option to delete the scandir of a given user some
directories of the scandir may not be removed if another user accesses them. In
our previous example where we created a scandir for root with UID *0* and a
regular user with the UID *1000* this would imply the following:

*%%livedir%%/scandir/0*		# will be deleted++
*%%livedir%%/tree/0*			# will be deleted++
*%%livedir%%/log/0*			# will be deleted++
*%%livedir%%/state/0*			# will be deleted

The _live_ directory of the *root* user (by default *%%livedir%%* is used) will
not be removed because another user (the one from our example) can still use
it. In fact *66-scandir* will only remove the subdirectories of the
corresponding UID of the _owner_ while the _live_ root directory is not
touched. If _live_ was created on a RAM filesystem the deletion happens on the
next reboot.

*Note*: A running scandir can not be removed. It needs to be stopped first with
	66-scanctl quit
or similar to be able to remove it.

# BOOT SPECIFICATION

The *-b* and *-s* options are called by *66-boot*(8) program. *-b* option
will create .s6-svscan control files (see *s6-svscan* interface documentation)
specifically for the _stage1_ a.k.a PID1 process. This special _scandir_ is
controlled by the safe wrappers *halt*, *poweroff*, *reboot*, *shutdown*
wrapper provided with *66* tools. The *66-shutdownd*(8) daemon which control the shutdown request will be created automatically at the correct location.
Further this specific tasks need to read of the _skeloton_ file _init.conf_
containing. the _live_ directory location which is the purpose of the *-s*
option.

The _live_ directory for a boot process requires writable directories, so in
order to accommodate read-only root filesystems, there needs to be a tmpfs
mounted before *s6-svscan* program can be run.


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 

src/66/66-scandir.c
/* 
 * 66-scandir.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 */

#include <string.h>
#include <errno.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
//#include <stdio.h>

#include <oblibs/obgetopt.h>
#include <oblibs/log.h>
#include <oblibs/directory.h>
#include <oblibs/types.h>
#include <oblibs/files.h>
#include <oblibs/string.h>
#include <oblibs/environ.h>

#include <skalibs/buffer.h>
#include <skalibs/stralloc.h>
#include <skalibs/djbunix.h>
#include <skalibs/types.h>
#include <skalibs/env.h>
#include <skalibs/bytestr.h>//byte_count

#include <s6/config.h>
#include <execline/config.h>

#include <66/config.h>
#include <66/utils.h>
#include <66/constants.h>
#include <66/environ.h>

#define CRASH 0
#define FINISH 1
#define HUP 2
#define INT 3
#define QUIT 4
#define TERM 5
#define USR1 6
#define USR2 7
#define PWR 8

#define SIGSIZE 64
#define AUTO_CRTE_CHW 1
#define AUTO_CRTE_CHW_CHM 2
#define PERM1777 S_ISVTX|S_IRWXU|S_IRWXG|S_IRWXO

static uid_t OWNER ;
static char OWNERSTR[UID_FMT] ;
static gid_t GIDOWNER ;
static char GIDSTR[GID_FMT] ;
static char TMPENV[MAXENV+1] ;
static char const *skel = SS_SKEL_DIR ;
static char const *log_user = SS_LOGGER_RUNNER ;
static unsigned int BOOT = 0 ;
unsigned int NOTIF = 0 ;

#define USAGE "66-scandir [ -h ] [ -z ] [ -v verbosity ] [ -b ] [ -l live ] [ -d notif ] [ -t rescan ] [ -L log_user ] [ -s skel ] [ -e environment ] [ -c | u | r ] owner"

static inline void info_help (void)
{
  static char const *help =
"66-scandir <options> owner\n"
"\n"
"options :\n"
"	-h: print this help\n"
"	-z: use color\n"
"	-v: increase/decrease verbosity\n"
"	-b: create scandir for a boot process\n"
"	-l: live directory\n"
"	-d: notify readiness on file descriptor\n"
"	-t: rescan scandir every milliseconds\n"
"	-L: run catch-all logger as log_user user\n"
"	-s: skeleton directory\n"
"	-e: directory environment\n"
"	-c: create scandir\n"
"	-r: remove scandir\n"
"	-u: bring up scandir\n"
;

	if (buffer_putsflush(buffer_1, help) < 0)
		log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
}

void scandir_up(char const *scandir, unsigned int timeout, char const *const *envp)
{
	int r ;
	r = scandir_ok(scandir) ;
	if (r < 0) log_dieusys(LOG_EXIT_SYS, "check: ", scandir) ;
	if (r)
	{
		log_trace("scandir: ",scandir," already running") ;
		return ;
	}
	
	unsigned int no = NOTIF ? 2 : 0 ;
	char const *newup[6 + no] ;
	unsigned int m = 0 ;
	char fmt[UINT_FMT] ;
	fmt[uint_fmt(fmt, timeout)] = 0 ;
	char notif[UINT_FMT] ;
	notif[uint_fmt(notif, NOTIF)] = 0 ;

	newup[m++] = S6_BINPREFIX "s6-svscan" ;
	if (no) {
		newup[m++] = "-d" ;
		newup[m++] = notif ;
	}
	newup[m++] = "-st" ;
	newup[m++] = fmt ;
	newup[m++] = "--" ;
	newup[m++] = scandir ;
	newup[m++] = 0 ;
	/** in case of boot process, the redirection of the uncaught-log
	 * will block the fd, so , do not sends anything
	 * VERBO2 strerr_warni3x("Starts scandir ",scandir," ...") ; */
	xpathexec_run (newup[0], newup, envp) ;
}

static void inline auto_chown(char const *str)
{
	log_trace("chown directory: ",str," to: ",OWNERSTR,":",GIDSTR) ;
	if (chown(str,OWNER,GIDOWNER) < 0)
		log_dieusys(LOG_EXIT_SYS,"chown: ",str) ;
}

static void inline auto_dir(char const *str,mode_t mode)
{
	log_trace("create directory: ",str) ;
	if (!dir_create(str,mode)) 
		log_dieusys(LOG_EXIT_SYS,"create directory: ",str) ;
}

static void inline auto_chmod(char const *str,mode_t mode)
{
	//VERBO3 strerr_warnt2x("chmod: ",str) ;
	if (chmod(str,mode) < 0)
		log_dieusys(LOG_EXIT_SYS,"chmod: ",str) ;
}

static void inline auto_file(char const *dst,char const *file,char const *contents,size_t conlen)
{
	log_trace("write file: ",dst,"/",file) ;
	if (!file_write_unsafe(dst,file,contents,conlen))
		log_dieusys(LOG_EXIT_SYS,"write file: ",dst,"/",file) ;
}

static void inline auto_stralloc(stralloc *sa,char const *str)
{
	if (!stralloc_cats(sa,str)) log_dieusys(LOG_EXIT_SYS,"append stralloc") ;
}

static void inline auto_check(char const *str,mode_t check,mode_t type,mode_t perm,int what)
{
	int r ;
	r = scan_mode(str,check) ;
	if (r < 0) { errno = EEXIST ; log_diesys(LOG_EXIT_SYS,"conflicting format of: ",str) ; }
	if (!r)
	{
		auto_dir(str,type) ;
		if (what > 0) auto_chown(str) ;
		if (what > 1) auto_chmod(str,perm) ;
	}
}

static void inline auto_fifo(char const *str)
{
	int r ;
	r = scan_mode(str,S_IFIFO) ;
	if (r < 0) { errno = EEXIST ; log_diesys(LOG_EXIT_SYS,"conflicting format of: ",str) ; }
	if (!r)
	{
		log_trace("create fifo: ",str) ;
		if (mkfifo(str, 0600) < 0) 
			log_dieusys(LOG_EXIT_SYS,"create fifo: ",str) ;
	}
}
static void inline auto_rm(char const *str)
{
	int r ;
	r = scan_mode(str,S_IFDIR) ;
	if (r > 0)
	{
		log_info("removing: ",str,"...") ;
		if (rm_rf(str) < 0) log_dieusys(LOG_EXIT_SYS,"remove: ",str) ;
	}
}
static void inline log_perm(char const *str,uid_t *uid,gid_t *gid)
{
	if (!youruid(uid,str)) log_dieusys(LOG_EXIT_SYS,"set uid of: ",str) ;
	if (!yourgid(gid,*uid)) log_dieusys(LOG_EXIT_SYS,"set gid of: ",str) ;
}

static void inline auto_addlive(stralloc *sa,char const *live, char const *str)
{
	auto_stralloc(sa,live) ;
	auto_stralloc(sa,str) ;
}

void write_shutdownd(char const *live, char const *scandir)
{
	stralloc run = STRALLOC_ZERO ;
	size_t scandirlen = strlen(scandir) ;
	char shut[scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN + 5 + 1] ;
	memcpy(shut,scandir,scandirlen) ;
	shut[scandirlen] = '/' ;
	memcpy(shut + scandirlen + 1,SS_BOOT_SHUTDOWND,SS_BOOT_SHUTDOWND_LEN) ;	
	shut[scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN] = 0 ;	
	auto_check(shut,S_IFDIR,0755,0755,AUTO_CRTE_CHW_CHM) ;
	memcpy(shut + scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN,"/fifo",5) ;
	shut[scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN + 5] = 0 ;
	auto_fifo(shut) ;
	auto_stralloc(&run, 
		"#!" EXECLINE_SHEBANGPREFIX "execlineb -P\n" \
		SS_BINPREFIX "66-shutdownd -l ") ;
	auto_stralloc(&run,live) ;
	auto_stralloc(&run," -s ") ;
	auto_stralloc(&run,skel) ;
	auto_stralloc(&run," -g 3000\n") ;
	shut[scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN] = 0 ;
	auto_file(shut,"run",run.s,run.len) ;
	memcpy(shut + scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN,"/run",4) ;
	shut[scandirlen + 1 + SS_BOOT_SHUTDOWND_LEN + 4] = 0 ;
	auto_chmod(shut,0755) ;
	stralloc_free(&run) ;
}

void write_bootlog(char const *live, char const *scandir) 
{
	int r ;
	uid_t uid ;
	gid_t gid ;
	size_t livelen = strlen(live) ;
	size_t scandirlen = strlen(scandir) ;
	size_t ownerlen = uid_fmt(OWNERSTR,OWNER) ;
 	size_t loglen ;
	char path[livelen + 4 + ownerlen + 1] ;
	char logdir[scandirlen + SS_SCANDIR_LEN + SS_LOG_SUFFIX_LEN + 1 + 5 + 1] ;
	
	stralloc run = STRALLOC_ZERO ;
	/** run/66/scandir/uid_name/scandir-log */
	memcpy(logdir,scandir,scandirlen) ;
	memcpy(logdir + scandirlen,"/" SS_SCANDIR SS_LOG_SUFFIX,SS_SCANDIR_LEN + SS_LOG_SUFFIX_LEN + 1) ;
	loglen = scandirlen + SS_SCANDIR_LEN + SS_LOG_SUFFIX_LEN + 1 ; 
	logdir[loglen] = 0 ;
	auto_check(logdir,S_IFDIR,0755,0,AUTO_CRTE_CHW) ;
	
	/** make the fifo*/
	memcpy(logdir + loglen, "/fifo", 5) ;
	logdir[loglen + 5] = 0 ;
	auto_fifo(logdir) ;
		
	/** set the log path for the run file
	 * /run/66/log*/
	memcpy(path,live,livelen) ;
	memcpy(path+livelen,"log",3) ;
	path[livelen + 3] = '/' ;
	memcpy(path + livelen + 4,OWNERSTR,ownerlen) ;
	path[livelen + 4 + ownerlen] = 0 ;
	log_trace("create directory: ",path) ;
	r = dir_create_parent(path,02750) ;
	if (!r) log_dieusys(LOG_EXIT_SYS,"create: ",path) ;
	log_perm(log_user,&uid,&gid) ;
	if (chown(path,uid,gid) < 0)
		log_dieusys(LOG_EXIT_SYS,"chown: ",path) ;
	auto_chmod(path,02755) ;

	/** make run file */
	auto_stralloc(&run, 
		"#!" EXECLINE_SHEBANGPREFIX "execlineb -P\n" \
		EXECLINE_BINPREFIX "redirfd -w 2 /dev/console\n" \
		EXECLINE_BINPREFIX "redirfd -w 1 /dev/null\n" \
		EXECLINE_BINPREFIX "redirfd -rnb 0 fifo\n" \
		S6_BINPREFIX "s6-setuidgid ") ;
	auto_stralloc(&run,log_user) ;
	auto_stralloc(&run,"\n" S6_BINPREFIX "s6-log -bpd3 -- 1 t ") ;
	auto_stralloc(&run,path) ;
	auto_stralloc(&run,"\n") ;
	
	logdir[loglen] = 0 ;
	
	auto_file(logdir,"run",run.s,run.len) ;
	auto_file(logdir,"notification-fd","3\n",2) ;
	
	memcpy(logdir + loglen,"/run",4) ;
	logdir[loglen + 4] = 0 ;	
	auto_chmod(logdir,0755) ;
	
	stralloc_free(&run) ;
}

void write_control(char const *scandir,char const *live, char const *filename, int file)
{
	size_t scandirlen = strlen(scandir) ;
	size_t filen = strlen(filename) ;
	char mode[scandirlen + SS_SVSCAN_LOG_LEN + filen + 1] ;
	stralloc sa = STRALLOC_ZERO ;

	/** shebang */
	auto_stralloc(&sa, "#!" EXECLINE_SHEBANGPREFIX "execlineb -P\n") ;
	
	if (file == FINISH)
	{
		if (BOOT)
		{
			auto_stralloc(&sa,
				EXECLINE_BINPREFIX "redirfd -w 1 /dev/console\n" \
				EXECLINE_BINPREFIX "fdmove -c 2 1\n" \
				EXECLINE_BINPREFIX "foreground { " SS_BINPREFIX "66-echo -- \"scandir ") ;
			auto_stralloc(&sa,scandir) ;
			auto_stralloc(&sa,
				" exited. Rebooting.\" }\n" \
				SS_BINPREFIX "66-hpr -r -f -l ") ;
			auto_addlive(&sa,live,"\n") ;
		}
		else
		{
			auto_stralloc(&sa,SS_BINPREFIX "66-echo -- \"scandir ") ;
			auto_stralloc(&sa, scandir) ;
			auto_stralloc(&sa, " shutted down...\"\n") ;
		}
		goto write ;
	}
	
	if (file == CRASH)
	{
		auto_stralloc(&sa,
			EXECLINE_BINPREFIX "redirfd -w 1 /dev/console\n" \
			EXECLINE_BINPREFIX "fdmove -c 2 1\n" \
			EXECLINE_BINPREFIX "foreground { " SS_BINPREFIX "66-echo -- \"scandir ") ;
		auto_stralloc(&sa,scandir) ;
		auto_stralloc(&sa, " crashed.") ;
		if (BOOT)
		{
			auto_stralloc(&sa,
				" Rebooting.\" }\n" \
				SS_BINPREFIX "66-hpr -r -f -l ") ;
			auto_addlive(&sa,live,"\n") ;
		}
		else auto_stralloc(&sa,"\" }\n") ;
		goto write ;
	}
	if (!BOOT)
	{
		auto_stralloc(&sa, EXECLINE_BINPREFIX "foreground { " SS_BINPREFIX "66-all -v3 -l ") ;
		auto_addlive(&sa,live," down }\n") ;
	}
	switch(file)
	{
		case PWR:
		case USR1:
			if (BOOT)
			{
				auto_stralloc(&sa,SS_BINPREFIX "66-shutdown -a -p -l ") ;
				auto_addlive(&sa,live," -- now\n") ;
			}
			break ;
		case USR2: 
			if (BOOT)
			{
				auto_stralloc(&sa,SS_BINPREFIX "66-shutdown -a -h -l ") ;
				auto_addlive(&sa,live," -- now\n") ;
			}
			break ;
		case TERM:
			if (!BOOT)
			{
				auto_stralloc(&sa, S6_BINPREFIX "66-scanctl -l ") ;
				auto_addlive(&sa,live," t\n") ;
			}
			break ;
		case QUIT:
			if (!BOOT) auto_stralloc(&sa, SS_BINPREFIX "66-scanctl quit\n") ;
			break ;
		case INT:
			if (BOOT)
			{
				auto_stralloc(&sa,SS_BINPREFIX "66-shutdown -a -r -l ") ;
				auto_addlive(&sa,live," -- now\n") ;
			}
			break ;
		case HUP:
			break ;
		default:
			break ;
	}
	
	write:
		memcpy(mode,scandir,scandirlen) ;
		memcpy(mode + scandirlen, SS_SVSCAN_LOG, SS_SVSCAN_LOG_LEN) ;
		mode[scandirlen + SS_SVSCAN_LOG_LEN ] = 0 ;
		auto_file(mode,filename+1,sa.s,sa.len) ;	
		memcpy(mode + scandirlen + SS_SVSCAN_LOG_LEN,filename,filen) ;
		mode[scandirlen + SS_SVSCAN_LOG_LEN + filen] = 0 ;
		auto_chmod(mode,0755) ;
			
	stralloc_free(&sa) ;
}

void create_scandir(char const *live, char const *scandir)
{
	size_t scanlen = strlen(scandir) ;
	char tmp[scanlen + 11 + 1] ;
	
	/** run/66/scandir/<uid> */	
	memcpy(tmp,scandir,scanlen) ;
	tmp[scanlen] = 0 ;
	auto_check(tmp,S_IFDIR,0755,0,AUTO_CRTE_CHW) ;
	
	/** run/66/scandir/name/.svscan */
	memcpy(tmp + scanlen, SS_SVSCAN_LOG, SS_SVSCAN_LOG_LEN) ;
	tmp[scanlen + SS_SVSCAN_LOG_LEN] = 0 ;
	auto_check(tmp,S_IFDIR,0755,0,AUTO_CRTE_CHW) ;
	
	char const *const file[] = 
	{ 
		"/crash", "/finish", "/SIGHUP", "/SIGINT",
		"/SIGQUIT", "/SIGTERM", "/SIGUSR1", "/SIGUSR2",
		"/SIGPWR"
	 } ;
	log_trace("write control file... ") ;
	for (int i = 0 ; i < 9; i++)
		write_control(scandir,live,file[i],i) ;

	if (BOOT)
	{
		write_bootlog(live, scandir) ;
		write_shutdownd(live, scandir) ;
	}
}

void sanitize_live(char const *live)
{
	size_t livelen = strlen(live) ;	
	char tmp[livelen + SS_SCANDIR_LEN + 1] ;
	
	/** run/66 */
	auto_check(live,S_IFDIR,0755,0,AUTO_CRTE_CHW) ;
	
	/** run/66/scandir */
	memcpy(tmp,live,livelen) ;
	memcpy(tmp + livelen,SS_SCANDIR,SS_SCANDIR_LEN) ;
	tmp[livelen + SS_SCANDIR_LEN] = 0 ;
	auto_check(tmp,S_IFDIR,0755,PERM1777,AUTO_CRTE_CHW_CHM) ;
		
	/** run/66/tree */
	memcpy(tmp + livelen,SS_TREE,SS_TREE_LEN) ;
	tmp[livelen + SS_TREE_LEN] = 0  ;
	auto_check(tmp,S_IFDIR,0755,PERM1777,AUTO_CRTE_CHW_CHM) ;

	/** run/66/log */
	memcpy(tmp + livelen,SS_LOG,SS_LOG_LEN) ;
	tmp[livelen + SS_LOG_LEN] = 0  ;
	auto_check(tmp,S_IFDIR,0755,PERM1777,AUTO_CRTE_CHW_CHM) ;
	
	/** /run/66/state*/
	memcpy(tmp + livelen,SS_STATE+1,SS_STATE_LEN-1) ;
	tmp[livelen + SS_STATE_LEN-1] = 0  ;
	auto_check(tmp,S_IFDIR,0755,PERM1777,AUTO_CRTE_CHW_CHM) ;
}

int main(int argc, char const *const *argv, char const *const *envp)
{
	int r ;
	unsigned int up, rescan, create, remove ;
	
	stralloc live = STRALLOC_ZERO ;
	stralloc scandir = STRALLOC_ZERO ;
	stralloc envdir = STRALLOC_ZERO ;
	
	char const *newenv[MAXENV+1] ;
	char const *const *genv = 0 ;
	
	log_color = &log_color_disable ;
	
	up = rescan = create = remove = 0 ;
	
	PROG = "66-scandir" ;
	{
		subgetopt_t l = SUBGETOPT_ZERO ;

		for (;;)
		{
			int opt = getopt_args(argc,argv, ">hzv:bl:d:t:s:e:cruL:", &l) ;
			if (opt == -1) break ;
			if (opt == -2) log_die(LOG_EXIT_USER,"options must be set first") ;
			switch (opt)
			{
				case 'h' : info_help(); return 0 ;
				case 'z' : log_color = !isatty(1) ? &log_color_disable : &log_color_enable ; break ;
				case 'v' : if (!uint0_scan(l.arg, &VERBOSITY)) log_usage(USAGE) ; break ;
				case 'b' : BOOT = 1 ; break ;
				case 'l' : if(!stralloc_cats(&live,l.arg)) log_die_nomem("stralloc") ;
						   if(!stralloc_0(&live)) log_die_nomem("stralloc") ;
						   break ;
				case 'd' : if (!uint0_scan(l.arg, &NOTIF)) log_usage(USAGE) ;
						   if (NOTIF < 3) log_die(LOG_EXIT_USER, "notification fd must be 3 or more") ;
						   if (fcntl(NOTIF, F_GETFD) < 0) log_diesys(LOG_EXIT_USER, "invalid notification fd") ;
						   break ;
				case 't' : if (!uint0_scan(l.arg, &rescan)) break ;
				case 's' : skel = l.arg ; break ;
				case 'e' : if(!stralloc_cats(&envdir,l.arg)) log_die_nomem("stralloc") ;
						   if(!stralloc_0(&envdir)) log_die_nomem("stralloc") ;
						   break ;
				case 'c' : create = 1 ; if (remove) log_usage(USAGE) ; break ;
				case 'r' : remove = 1 ; if (create) log_usage(USAGE) ; break ;
				case 'u' : up = 1 ; if (remove) log_usage(USAGE) ; break ;
				case 'L' : log_user = l.arg ; break ;
				default :  log_usage(USAGE) ; 
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	
	if (!argc) OWNER = MYUID ;
	else if (!youruid(&OWNER,argv[0])) log_dieusys(LOG_EXIT_SYS,"set uid of: ",argv[0]) ;
		
	if (BOOT && NOTIF) NOTIF = 0 ;
	if (BOOT && OWNER) log_die(LOG_EXIT_USER,"-b options can be set only with root") ; 
	OWNERSTR[uid_fmt(OWNERSTR,OWNER)] = 0 ;
	
	if (!yourgid(&GIDOWNER,OWNER)) log_dieusys(LOG_EXIT_SYS,"set gid of: ",OWNERSTR) ;
	GIDSTR[gid_fmt(GIDSTR,GIDOWNER)] = 0 ; 
		
	/** live -> /run/66/ */
	r = set_livedir(&live) ;
	if (r < 0) log_die(LOG_EXIT_USER,"live: ",live.s," must be an absolute path") ;
	if (!r) log_dieusys(LOG_EXIT_SYS,"set live directory") ;
	
	if (!stralloc_copy(&scandir,&live)) log_die_nomem("stralloc") ;
		
	/** scandir -> /run/66/scandir/ */
	r = set_livescan(&scandir,OWNER) ;
	if (r < 0) log_die(LOG_EXIT_USER,"scandir: ", scandir.s, " must be an absolute path") ;
	if (!r) log_dieusys(LOG_EXIT_SYS,"set scandir directory") ;
		
	if (BOOT && skel[0] != '/')
		log_die(LOG_EXIT_USER, "rc.shutdown: ",skel," must be an absolute path") ;
	
	if (envdir.len)
	{
		if (envdir.s[0] != '/')
			log_die(LOG_EXIT_USER,"environment: ",envdir.s," must be an absolute path") ;
		
		if (!environ_get_envfile_n_merge(envdir.s,envp,newenv,TMPENV)) log_dieusys(LOG_EXIT_SYS,"build environment with: ",envdir.s) ;
		genv = newenv ;
	}
	else genv = envp ;
	
	r = scan_mode(scandir.s, S_IFDIR) ;
	if (r < 0) log_die(LOG_EXIT_SYS,"scandir: ",scandir.s," exist with unkown mode") ;
	if (!r && !create && !remove) log_die(LOG_EXIT_USER,"scandir: ",scandir.s," doesn't exist") ;
	if (!r && create)
	{
		log_trace("sanitize ",live.s," ...") ;
		sanitize_live(live.s) ;
		log_trace("create scandir ",scandir.s," ...") ;
		create_scandir(live.s, scandir.s) ;
	}
	/**swap to char [] to be able to freed stralloc*/
	char ownerscan[scandir.len + 1] ;
	memcpy(ownerscan,scandir.s,scandir.len) ;
	ownerscan[scandir.len] = 0 ;
	if (r && create)
	{
		log_info("scandir: ",scandir.s," already exist, keep it") ;
		goto end ;
	}
 	
 	r = scandir_ok(scandir.s) ;
	if (r < 0) log_dieusys(LOG_EXIT_SYS, "check: ", scandir.s) ;
	if (r && remove) log_dieu(LOG_EXIT_USER,"remove: ",scandir.s,": is running")  ;
	if (remove)
	{
		auto_rm(scandir.s) ;
		/** /run/66/tree/uid */
		if (!stralloc_copy(&scandir,&live)) log_die_nomem("stralloc") ;
		r = set_livetree(&scandir,OWNER) ;
		if (!r) log_dieusys(LOG_EXIT_SYS,"set livetree directory") ;
		auto_rm(scandir.s) ;
		if (!stralloc_copy(&scandir,&live)) log_die_nomem("stralloc") ;
		/** run/66/state/uid */
		r = set_livestate(&scandir,OWNER) ;
		if (!r) log_dieusys(LOG_EXIT_SYS,"set livestate directory") ;
		auto_rm(scandir.s) ;
	}
	end:
	stralloc_free(&scandir) ;
	stralloc_free(&live) ;
	stralloc_free(&envdir) ;
	
	if (up) scandir_up(ownerscan,rescan,genv) ;
		
	return 0 ;
}
